src/tactic/interactive.lean

```diff
@@ -937,12 +937,10 @@ end
 meta def set (h_simp : parse (tk "!")?) (a : parse ident) (tp : parse ((tk ":") >> texpr)?)
   (_ : parse (tk ":=")) (pv : parse texpr)
   (rev_name : parse opt_dir_with) :=
-do let vt := match tp with | some t := t | none := pexpr.mk_placeholder end,
-   let pv := ``(%%pv : %%vt),
-   v ← to_expr pv,
-   tp ← infer_type v,
-   definev a tp v,
-   when h_simp.is_none $ change' pv (some (expr.const a [])) loc.wildcard,
+ do let vt := tp.get_or_else pexpr.mk_placeholder,
+   tp ← i_to_expr vt,
+   to_expr ``(%%pv : %%tp) >>= definev a tp,
+   when h_simp.is_none $ change' ``(%%pv : %%vt) (some (expr.const a [])) loc.wildcard,
```

Comment:
This elaborates `pv` and `vt` twice, right?

Reply:
It's subtle. Changing at a hypothesis requires reverting and reintroducing it. If you elaborate `pv` and then naively try to change `pv` at *, you can end up reverting and reintroducing variables that `pv` depends on; if you do this, `pv` is no longer a valid expression, since those variables have changed.

Possible solutions:
* Instead of trying to change at *, change only at hypotheses that `pv` doesn't depend on. Do we have a function to compute this? `kdepends_on` isn't enough, you need more than a syntactic check. 
* Change the `change at` logic so that when it doesn't make any changes, nothing gets reverted. I haven't looked into this and I'm not sure if there are complications.

This is more work than I was expecting for a quick PR! (Note that it was already elaborating `pv` twice before.)

Reply:
> * Change the `change at` logic so that when it doesn't make any changes, nothing gets reverted. I haven't looked into this and I'm not sure if there are complications.

Actually, I think this is straightforward.